pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go

```diff
@@ -0,0 +1,1339 @@
+	// which is messy.
+	Entries() []raftpb.Entry
+	// RetransmitMsgApps returns the MsgApps that are being retransmitted, or
+	// being used to ping the follower. These will never be queued by
+	// replication AC, but it may adjust the priority before sending (due to
+	// priority inheritance, when the Message has non-empty Entries).
+	RetransmitMsgApps() []raftpb.Message
+}
+
+// RaftInterface abstracts what the RangeController needs from the raft
+// package, when running at the leader. It also provides one piece of
+// information that is not in the raft package -- the current connected state
+// of the store, that we will use the RaftTransport to answer.
+//
+// NB: group membership and connectivity information is communicated to the
+// RangeController via a separate channel, as in the v1 implementation, since
+// it depends on information inside and outside Raft. The methods in Raft are
+// the ones RangeController will call.
+//
+// The implementation should not need to acquire Replica.mu since we have not
+// said anything about the relative lock ordering of Replica.mu and the
+// internal mutexes in RangeController.
+type RaftInterface interface {
+	// FollowerState returns the current state of a follower. The value of
+	// nextUpperBound is populated iff in StateReplicate. All entries >=
+	// nextUpperBound have not yet had MsgApps constructed.
+	//
+	// When a follower transitions from {StateProbe,StateSnapshot} =>
+	// StateReplicate, or was already in StateReplicate but was disconnected and
+	// has now reconnected, we can now start trying to send MsgApps. We should
+	// notice such transitions both in HandleRaftEvent and SetReplicas. We
+	// should *not* construct MsgApps for a StateReplicate follower that is
+	// disconnected -- there is no timeliness guarantee on how long a follower
+	// will stay in StateReplicate despite it being down, and by sending such a
+	// follower MsgApps that are not being received we are defeating flow
+	// control (since we will have advanced nextUpperBound).
+	//
+	// RACv1 also cared about three other cases where the follower behaved as if
+	// it were disconnected (a) paused follower, (b) follower is behind, (c)
+	// follower is inactive (see
+	// replicaFlowControlIntegrationImpl.notActivelyReplicatingTo). (b) and (c)
+	// were needed since it paced at rate of slowest replica, while for regular
+	// work we will in v2 pace at slowest in quorum (and we don't care about
+	// elastic experiencing a hiccup, given it paces at rate of slowest). For
+	// (a), we plan to remove follower pausing. So the v2 code will be
+	// simplified.
+	FollowerState(replicaID roachpb.ReplicaID) (state tracker.StateType, nextUpperBound uint64)
+	FollowerTransportConnected(storeID roachpb.StoreID) bool
+	// HighestEntryIndex is the highest index assigned in the log, and produced
+	// in Ready.Entries(). If there have been no entries produced, since this
+	// replica became the leader, this is the commit index.
```

Comment:
A newly elected leader can have a suffix of entries which are not yet committed. Should we initialize to the log size, rather than to the commit index?

A thing about the commit index on the new leader: it doesn't know it before it commits a dummy entry at its term. So, by the time the leader learns the commit index, it has already appended an entry to its log.

Do we need to tie this definition to `Ready.Entries()`? We can get the size of the log more directly, by accessing `raft.raftLog.lastEntryID()`. Would it suffice to say that this returns the last index in the log at all times, and not mention the "becoming leader" case?